### PR TITLE
cpu/cortexm_common: Flush pipeline after triggering PendSV

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Flush pipeline after PendSV in `thread_yield_higher()`. See https://interrupt.memfault.com/blog/arm-cortex-m-exceptions-and-nvic#pendsv-example

@kaspar030: Thanks for pointing out the fix.

### Testing procedure

See "steps to reproduce" in https://github.com/RIOT-OS/RIOT/issues/15085

### Issues/PRs references

Fixes https://github.com/RIOT-OS/RIOT/issues/15085

### CODE REVIEW

File: cpu/cortexm_common/thread_arch.c [hunk 1]
```diff
@@ -395,6 +397,7 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
     "ldr    r0, [r0]                  \n" /* dereference TCB */
     "ldr    r0, [r0]                  \n" /* load tcb-sp to R0 */
     "mov    sp, r0                    \n" /* make user mode SP active SP */
+    "ibs                              \n" /* flush pipeline */
```
Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c:400]: I think this is only necessary when using msr, not by writing to sp in general. (gcc generates tons of sp modifying code in the function prologues, I've not seen a pipeline flush in there).

File: cpu/cortexm_common/thread_arch.c [hunk 2]
```diff
@@ -372,6 +373,7 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
 
     "mrs    r0, psp                   \n" /* get stack pointer from user mode */
     "mov    sp, r0                    \n" /* set user mode SP as active SP */
+    "ibs                              \n" /* flush pipeline */
```
Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c:376]: not needed, see other comment

File: cpu/cortexm_common/thread_arch.c [hunk 3]
```diff
@@ -405,6 +408,7 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
     "mov    r1, sp                    \n" /* restore the user mode SP */
     "msr    psp, r1                   \n" /* for this write it to the PSP reg */
     "mov    sp, r12                    \n" /* and get the parked MSR SP back */
+    "ibs                              \n" /* flush pipeline */
```
Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c:411]: here I'm not sure. the write to sp doesn't need this, the msr instruction above might. but as the psp is only used after exception return, I don't think it is really needed.

**Review by kaspar030 — CHANGES_REQUESTED** [review]

**Comment by kaspar030** [comment]
>  this fixes #15085

but, the __dsb() is the fix for #15085, right? (I don't think they should be in the same PR)

**Comment by maribu** [comment]
> but, the __dsb() is the fix for #15085, right?

Yes. But I think it should have been an ISB there as well, at least according to some guy in the internet. I can confirm that the issue is seemingly gone with ISB as well.

> (I don't think they should be in the same PR)

Dropped the other isb

**Review by kaspar030 — APPROVED** [review]
ACK.

While I'm not sure this is necessary for Cortex-M0, and I'm not sure it is necessary if thread_yield_higher() is a regular function on M3/M4, I'm willing to accept the extra instruction just to avoid surprises when inlining or using LTO.

And it certainly fixes issues on M7.

**Comment by maribu** [comment]
Ignoring all ESP failures, the following tests failed:

- [tests/pkg_libhydrogen](https://ci.riot-os.org/RIOT-OS/RIOT/15086/49f3d1056d7e7a973edf8125f82232a4576588c6/output/run_test/tests/pkg_libhydrogen/samr21-xpro:llvm.txt) on `samr21-xpro` when compiled with LLVM
- [tests/pkg_utensor](https://ci.riot-os.org/RIOT-OS/RIOT/15086/49f3d1056d7e7a973edf8125f82232a4576588c6/output/run_test/tests/pkg_utensor/samr21-xpro:llvm.txt) with `samre-xpro` when compiled with LLVM
- [tests/pkg_lvgl](https://ci.riot-os.org/RIOT-OS/RIOT/15086/49f3d1056d7e7a973edf8125f82232a4576588c6/output/compile/tests/pkg_lvgl/wemos-zero:gnu.txt) on `wemos-zero` when compiled with GCC

I can investigate the last one, but I have not ~~Atmel~~Microchip Xpro at hand. (I consider them too expensive...)

**Comment by aabadie** [comment]
> tests/pkg_utensor with samre-xpro when compiled with LLVM

This one should be fixed by #15089

**Comment by maribu** [comment]
>     * [tests/pkg_lvgl](https://ci.riot-os.org/RIOT-OS/RIOT/15086/49f3d1056d7e7a973edf8125f82232a4576588c6/output/compile/tests/pkg_lvgl/wemos-zero:gnu.txt) on `wemos-zero` when compiled with GCC

Here the culprit are connectivity issues on the worker, cloning `lvgl` failed.

Nice to know that the board files are indeed used :-)

**Comment by maribu** [comment]
On my `wemos-zero` (`samd21`, not `samr21`), `tests/pkg_libhydrogen` is working. Maybe that was just a glitch?

**Comment by bergzand** [comment]
> On my `wemos-zero` (`samd21`, not `samr21`), `tests/pkg_libhydrogen` is working. Maybe that was just a glitch?

It's not working on my samr21-xpro: https://github.com/RIOT-OS/RIOT/issues/15066 :cry:

**Comment by bergzand** [comment]
But it is unrelated to this PR

**Comment by maribu** [comment]
OK, with all tests failures due to other reasons, I'll rerun murdock without tests.

I wonder if in such a situation I would be allowed to disable compile checks (as those have passed an hour ago already).

**Comment by aabadie** [comment]
> I would be allowed to disable compile checks

Even if it's not a super nice practice, some people already do that in such situations, to save a bit of CPU time.

**Comment by maribu** [comment]
All green :-)

**Comment by maribu** [comment]
Thanks :-)

**Pull request merged**